### PR TITLE
HBASE-23347 Allow custom authentication methods for RPCs; addendum

### PR DESCRIPTION
**Pull request by petersomogyi**
Fix error prone issues

### CODE REVIEW

**Comment by joshelser** [comment]
Looks great! Thanks for fixing, @petersomogyi . I didn't realize that errorprone wasn't run during QA!

One thing - MetaBrowser wasn't changed by my commit. Looks like that might have come in via @ndimiduk 's HBASE-23653?

**Comment by petersomogyi** [comment]
> One thing - MetaBrowser wasn't changed by my commit. Looks like that might have come in via @ndimiduk 's HBASE-23653?

I haven't noticed it was from a different commit. Let me remove that modification here and create an addendum for HBASE-23653 for easier tracking.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 33s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  1s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -0 :warning: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 37s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   5m 36s |  master passed  |
| +1 :green_heart: |  compile  |   1m 31s |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 56s |  master passed  |
| +1 :green_heart: |  shadedjars  |   4m 49s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  7s |  master passed  |
| +0 :ok: |  spotbugs  |   5m  4s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   6m 15s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 17s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   5m  9s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 35s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 35s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 55s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   4m 54s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |  16m 19s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| +1 :green_heart: |  javadoc  |   1m  6s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   6m 34s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |   1m 59s |  hbase-client in the patch failed.  |
| -1 :x: |  unit  | 172m 18s |  hbase-server in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 57s |  The patch does not generate ASF License warnings.  |
|  |   | 242m 17s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.security.provider.TestSaslClientAuthenticationProviders |
|   | hadoop.hbase.master.assignment.TestRegionMoveAndAbandon |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1060/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1060 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux b3ad105a52a9 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1060/out/precommit/personality/provided.sh |
| git revision | master / c4395b52c9 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1060/2/artifact/out/patch-unit-hbase-client.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1060/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1060/2/testReport/ |
| Max. process+thread count | 5216 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1060/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

File: hbase-client/src/main/java/org/apache/hadoop/hbase/security/provider/SaslClientAuthenticationProviders.java [hunk 1]
```diff
@@ -148,8 +150,9 @@ static void addExplicitProviders(Configuration conf,
       // Instantiate it
       SaslClientAuthenticationProvider provider;
       try {
-        provider = (SaslClientAuthenticationProvider) clz.newInstance();
-      } catch (InstantiationException | IllegalAccessException e) {
+        provider = (SaslClientAuthenticationProvider) clz.getConstructor().newInstance();
```
Comment (petersomogyi) [inline hbase-client/src/main/java/org/apache/hadoop/hbase/security/provider/SaslClientAuthenticationProviders.java]: This modification causes TestSaslClientAuthenticationProviders#testDifferentConflictingImplementationsFail test failure. I'll look into it later.

**Comment by petersomogyi** [comment]
The test fails to instantiate the 2 provider classes:
`java.lang.NoSuchMethodException: org.apache.hadoop.hbase.security.provider.TestSaslClientAuthenticationProviders$ConflictingProvider1.<init>()`

I found a [JDK bug report](https://bugs.openjdk.java.net/browse/JDK-8189233) with the same symptoms where it is described in the last comment why this is the correct behavior.

tl;dr The default constructor won't be public for these classes so we either need to make the 2 classes public or explicitly add a public constructor.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 34s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 37s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   5m 29s |  master passed  |
| +1 :green_heart: |  compile  |   1m 29s |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 40s |  master passed  |
| +1 :green_heart: |  shadedjars  |   4m 48s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  5s |  master passed  |
| +0 :ok: |  spotbugs  |   4m 58s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   6m  8s |  master passed  |
| -0 :warning: |  patch  |   5m 14s |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   5m  9s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 29s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 29s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 39s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   4m 50s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |  16m 17s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| +1 :green_heart: |  javadoc  |   1m  5s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   6m 24s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 10s |  hbase-client in the patch passed.  |
| +1 :green_heart: |  unit  | 165m 15s |  hbase-server in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  6s |  The patch does not generate ASF License warnings.  |
|  |   | 234m 28s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1060/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1060 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux baf291b40369 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1060/out/precommit/personality/provided.sh |
| git revision | master / 569ac1232a |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1060/3/testReport/ |
| Max. process+thread count | 5488 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1060/3/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.
